scipy/sparse/linalg/_dsolve/linsolve.py

```diff
@@ -651,93 +649,94 @@ def spsolve_triangular(A, b, lower=True, overwrite_A=False, overwrite_b=False,
-    if A.shape[0] != b.shape[0]:
+    if M != b.shape[0]:
         raise ValueError(
             'The size of the dimensions of A must be equal to '
             'the size of the first dimension of b but the shape of A is '
             '{} and the shape of b is {}.'.format(A.shape, b.shape))
 
-    # Init x as (a copy of) b.
-    x_dtype = np.result_type(A.data, b, np.float64)
-    if overwrite_b:
-        if np.can_cast(b.dtype, x_dtype, casting='same_kind'):
-            x = b
-        else:
-            raise ValueError(
-                'Cannot overwrite b (dtype {}) with result '
-                'of type {}.'.format(b.dtype, x_dtype))
-    else:
-        x = b.astype(x_dtype, copy=True)
+    A = A._asfptype()  # upcast to a floating point format
+    result_dtype = np.promote_types(A.dtype, b.dtype)
+    if A.dtype != result_dtype:
+        A = A.astype(result_dtype)
+    if b.dtype != result_dtype:
+        b = b.astype(result_dtype)
+    if not overwrite_b:
+        b = b.copy()
 
-    # Choose forward or backward order.
     if lower:
-        row_indices = range(len(b))
+        L = A
+        U = csc_matrix((N, N), dtype=result_dtype)
     else:
-        row_indices = range(len(b) - 1, -1, -1)
-
-    # Fill x iteratively.
-    for i in row_indices:
-
-        # Get indices for i-th row.
-        indptr_start = A.indptr[i]
-        indptr_stop = A.indptr[i + 1]
-
-        if lower:
-            A_diagonal_index_row_i = indptr_stop - 1
-            A_off_diagonal_indices_row_i = slice(indptr_start, indptr_stop - 1)
-        else:
-            A_diagonal_index_row_i = indptr_start
-            A_off_diagonal_indices_row_i = slice(indptr_start + 1, indptr_stop)
-
-        # Check regularity and triangularity of A.
-        if not unit_diagonal and (indptr_stop <= indptr_start
-                                  or A.indices[A_diagonal_index_row_i] < i):
-            raise LinAlgError(
-                f'A is singular: diagonal {i} is zero.')
-        if not unit_diagonal and A.indices[A_diagonal_index_row_i] > i:
-            raise LinAlgError(
-                'A is not triangular: A[{}, {}] is nonzero.'
-                ''.format(i, A.indices[A_diagonal_index_row_i]))
-
-        # Incorporate off-diagonal entries.
-        A_column_indices_in_row_i = A.indices[A_off_diagonal_indices_row_i]
-        A_values_in_row_i = A.data[A_off_diagonal_indices_row_i]
-        x[i] -= np.dot(x[A_column_indices_in_row_i].T, A_values_in_row_i)
-
-        # Compute i-th entry of x.
-        if not unit_diagonal:
-            x[i] /= A.data[A_diagonal_index_row_i]
+        L = eye(N, dtype=result_dtype, format='csc')
+        U = A
+        U.setdiag(0)
+
+    x, info = _superlu.gstrs(trans,
+                             N, L.nnz, L.data, L.indices, L.indptr,
+                             N, U.nnz, U.data, U.indices, U.indptr,
+                             b)
+    if info:
+        warn("Matrix is singular", MatrixRankWarning)
+        x.fill(np.nan)
```

Comment:
Did we not use to raise a `LinAlgError` if the matrix was singular?

Reply:
Thanks for the hint, this might be a remnant from an [earlier version of `spsolve`](https://github.com/scipy/scipy/blob/9cc8d29a21b8e02684ea11790d0a217fda84459a/scipy/sparse/linalg/_dsolve/linsolve.py#L275-L277), I fixed it in 34af5470a5aafbcfd8c1d8af42c24e1e3ff0c5c8.